README.md
### IDBac Easy Install

Find installers for Windows at: https://chasemc.github.io/IDBac/

### Finding Help, Reporting Issues


We have setup a discussion forum at https://groups.google.com/forum/#!forum/idbac

Issues/Errors may be reported here (https://github.com/chasemc/IDBacApp/issues) 


### Run IDBac from R

If you would rather run IDBac directly from R:

```{r}
install.packages("remotes") # Run this if you don't have the devtools package
remotes::install_github("chasemc/IDBacApp@*release")

# And, to run the app:
IDBacApp::run_app()

```


### IDBac Method

[Click Here](https://github.com/chasemc/IDBacApp/blob/master/method.md#-idbac-instruction-manual-for-maldi-tof-ms-acquistion-and-data-analysis-)


### Citing IDBac

Clark, C. M., Costa, M. S., Sanchez, L. M., & Murphy, B. T. (2018). Coupling MALDI-TOF mass spectrometry protein and specialized metabolite analyses to rapidly discriminate bacterial function. Proceedings of the National Academy of Sciences of the United States of America, 115(19), 4981-4986. [https://www.doi.org/10.1073/pnas.1801247115](https://www.doi.org/10.1073/pnas.1801247115)

Every IDBac release is preserved at Zenodo for citability/reproducibility:   
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1115619.svg)](https://doi.org/10.5281/zenodo.1115619)


### For IDBac Developers

Linux and Mac CI:
  - [![Build Status](https://travis-ci.org/chasemc/IDBacApp.svg?branch=master)](https://travis-ci.org/chasemc/IDBacApp)
  
Windows CI:
  - [![AppVeyor build status](https://ci.appveyor.com/api/projects/status/78sr4bqpncnwt52y/branch/master?svg=true)](https://ci.appveyor.com/project/chasemc/IDBacApp)


Coverage:
  - [![codecov](https://codecov.io/gh/chasemc/IDBacApp/branch/master/graph/badge.svg)](https://codecov.io/gh/chasemc/IDBacApp)


NEWS.md
---
title: "NEWS"
author: "Chase Clark"
---

Updates are ordered newest to oldest.


### Updated SQLite database table and field names

- This was done for consistency, especially to make it easier for others to program with IDBac databases

- Table Names:
  - "IndividualSpectra is now spectra"
    - "spectrumMassHash is now spectrum_mass_hash"
    - "spectrumIntensityHash is now spectrum_intensity_hash"
    - "XMLHash is now xml_hash"
    - "Strain_ID is now strain_id"
    - "peakMatrix is now peak_matrix"
    - "spectrumIntensity is now spectrum_intensity"
    - "maxMass is now max_mass"
    - "minMass is now min_mass"
    - "ignore is now ignore"
    - "number is now number"
    - "timeDelay is now time_delay"
    - "timeDelta is now time_delta"
    - "calibrationConstants is now calibration_constants"
    - "v1tofCalibration is now v1_tof_calibration"
    - "dataType is now data_type"
    - "dataSystem is now data_system"
    - "spectrometerType is now spectrometer_type"
    - "inlet is now inlet"
    - "ionizationMode is now ionization_mode"
    - "acquisitionMethod is now acquisition_method"
    - "acquisitionDate is now acquisition_date"
    - "acquisitionMode is now acquisition_mode"
    - "tofMode is now tof_mode"
    - "acquisitionOperatorMode is now acquisition_operator_mode"
    - "laserAttenuation is now laser_attenuation"
    - "digitizerType is now digitizer_type"
    - "flexControlVersion is now flex_control_version"
    - "id is now id"
    - "instrument is now instrument"
    - "instrumentId is now instrument_id"
    - "instrumentType is now instrument_type"
    - "massError is now mass_error"
    - "laserShots is now laser_shots"
    - "patch is now patch"
    - "path is now path"
    - "laserRepetition is now laser_repetition"
    - "spot is now spot"
    - "spectrumType is now spectrum_type"
    - "targetCount is now target_count"
    - "targetIdString is now target_id_string"
    - "targetSerialNumber is now target_serial_number"
    - "targetTypeNumber is now target_type_number"    
  - "XML is now xml"
    - "XMLHash is now xml_hash"
    - "XML is now xml"
    - "manufacturer is now manufacturer"
    - "model is now model"
    - "ionization is now ionization"
    - "analyzer is now analyzer"
    - "detector is now detector"
    - "Instrument_MetaFile is now instrument_metafile"
  - "locale is now locale"
    - "locale is now locale"
  - "massTable is now mass_index"
    - "spectrumMassHash is now spectrum_mass_hash"
    - "massVector is now mass_vector"
  - "metaData is now metadata"
    - "Strain_ID is now strain_id"
    - "Genbank_Accession is now genbank_accession"
    - "NCBI_TaxID is now ncbi_taxid"
    - "Kingdom is now kingdom"
    - "Phylum is now phylum"
    - "Class is now class"
    - "Order is now order"
    - "Family is now family"
    - "Genus is now genus"
    - "Species is now species"
    - "MALDI_Matrix is now maldi_matrix"
    - "DSM_Agar_Media is now dsm_cultivation_media"
    - "Cultivation_Temp_Celsius is now cultivation_temp_celsius"
    - "Cultivation_Time_Days is now cultivation_time_days"
    - "Cultivation_Other is now cultivation_other"
    - "User is now user_firstname_lastname"
    - "User_ORCID is now user_orcid"
    - "PI_FirstName_LastName is now pi_firstname_lastname"
    - "PI_ORCID is now pi_orcid"
    - "dna_16S is now dna_16s"  
  - "version is now version"
    - "IDBacVersion is now idbac_version"
    - "rVersion is now r_version"





## 1.1.10

- BugFix:
  -  When updating IDBac software, experiments saved into the default directory would sometimes be deleted by the installer. 
          This has been changed so that IDBac will try to make the default save directory '~Documents/IDBac_experiments'.
- Improvement:
  - Improved support for importing data from Microtyper instruments. Including 
- Improvement:
  - Changed how IDBac first connects to an experiment database, for faster connect time, especially with large experiments (1000's of samples).
- Improvement:
  - Mirror plot code was improved and placed into a standalone function for coding with IDBac outside of the Shiny app.
- BugFix:
  - 'Binning' algorithm checks for empty peak lists and added check for list output from `mapply()`.
- Improvement:  
  - Continuous integration fixed for Mac and Linux builds.



## 1.0.0

This release is more or less an entire re-write of IDBac, so... many, many changes. Only some are listed below.

### Installer Changes

RInno had a couple pretty massive API changes which led to my needing to write https://github.com/chasemc/electricShine.  This means that IDBac is now an Electron app that is self-contained (includes R, packages,Electron, browser, etc). All packages and R were installed from a specified MRAN snapshot. The installer is built using Continuous Deployment on appVeyor (https://github.com/chasemc/electricShine/blob/master/appveyor.yml) and attached to each GitHub release. 

### Structural Changes
This is likely the largest change and the most significant for anyone wanting to add-on or contribute to the codebase. 

While the first version of IDBac followed a valid paradigm for small Shiny apps (two files: server.r and ui.r), it became quickly apparent that it was an unsustainable and unscalable model for future development when shiny modules was released.

 New paradigms:
 - IDBac is an R package
   - Now IDBac is structured as a regular R-package. This means that significant computation and code chunks have been pulled out of server.r and are now located in the package's `R` directory. 
 - IDBac is testable
   - While methods for testing the interactivity of Shiny apps is still a little shaky, many functions have been extracted from the shiny logic and now have `testthat` tests, writing tests for old functions is an ongoing effort.
     - To contribute to writing tests, see https://codecov.io/gh/chasemc/IDBacApp for a breakdown of functions that need tests.
 
### Structural Changes (continued)


R/app_ui.R

#' Main UI of IDBac
#'
#' @return Main UI of IDBac
#' @export
#'
app_ui <- function(){
  
  #----------
  # Changes Zenodo DOI
  #zenodoId <- "1203781"
  
  # The UI section of the Shiny app provides the "User Interface" and is the means of user interaction.
  # "h3", "p", and "br" are HTML,the other lines setup the different panels, tabs, plots, etc of the user interface.
  #tags$body(background = "rotatingPCA.gif",
  navbarPage(collapsible = T,
             # theme = "css.css",
             title = "IDBac",
             id = "mainIDBacNav",
             tabPanel(
               title = "Introduction",
               div(class = "introPageImage",
                   
                   div(class = "introPage",
                       #includeCSS(file.path(system.file('www', package = 'IDBacApp'), 'style.css')),
                       tags$head(tags$link(href = "www/styles.css", rel = "stylesheet")),
                       tags$head(tags$link(href = "www/collapse/shinyBS.js")),
                       tags$head(tags$link(href = "www/collapse/shinyBS.css")),
                    #   shinyjs::useShinyjs(),
                                wellPanel(class = "intro_WellPanel",
                                          h1("Welcome to IDBac",
                                             align = "center"),
                                          br(),
                                          h4("General Information:",
                                             align = "center"),
                                          tags$ul(
                                            tags$li(
                                              p("For more information about the method, as well as links to the full code, please visit ", 
                                                a(href = "https://chasemc.github.io/IDBac/",
                                                  target = "_blank", "chasemc.github.io/IDBac")
                                              )
                                            ),
                                            tags$li(
                                              p("Bugs and suggestions may be reported on the ",
                                                a(href = "https://github.com/chasemc/IDBacApp/issues",
                                                  target = "_blank","IDBac issues page on GitHub", 
                                                  img(border = "0",
                                                      title = "https://github.com/chasemc/IDBacApp/issues",
                                                      src = "www/GitHub.png",
                                                      width = "25",
                                                      height = "25")
                                                )
                                              )
                                            ),
                                            tags$li(
                                              p("For discussions and general questions please use the ",
                                                a(href = "https://groups.google.com/forum/#!forum/idbac",
                                                  target = "_blank","IDBac Google group forum"
                                                )
                                              )
                                            ),
                                            column(width = 12,
                                                   align = "center",
                                                   actionButton("updateIDBac",
                                                                label = "Check for Updates",
                                                                align = "center")
                                            )
                                          ),
                                          br(),
                                          br(),
                                          tags$ul(
                                            tags$li(
                                              p("Publication:")),
                                            tags$ul(
                                              tags$li(
                                                
                                                div(id = 'pnasTitle',
                                                    class = 'publicationTitle',
                                                    "Coupling MALDI-TOF mass spectrometry protein and specialized metabolite analyses to rapidly discriminate bacterial function"),
                                                div(id = 'pnasauthors',
                                                    class = 'publicationAuthors',
                                                    "Chase M. Clark, Maria S. Costa, Laura M. Sanchez, Brian T. Murphy"),
                                                div(id = 'pnasDetails',
                                                    class = 'publicationDetails',
                                                    "PNAS May 8, 2018 115 (19) 4981-4986; doi:",
                                                    a(href = "https://doi.org/10.1073/pnas.1801247115", 
                                                      target = "_blank", "10.1073/pnas.1801247115")
                                                )
                                              )
                                            )
                                          ),
                                          tags$ul(
                                            tags$li(
                                              p("Software:")
                                            ),
                                            tags$ul(
                                              tags$li(
                                                p("For reproducibility, cite this specific version of IDBac as version:",
                                                  as.character(packageVersion("IDBacApp"))
                                                )
                                              )
                                            )
                                          ),
                                          br(),
                                          h4("Use the navigation bar at the top of the page to begin", 
                                             align = "center"),
                                          div(align = "center",
                                              IDBacApp::selectDirectory_UI("userWorkingDirectory",
                                                                           label = "Select IDBac Data Storage Location"),
                                              verbatimTextOutput("userWorkingDirectoryText")
                                          )
                                )
                   )
                       
                   
               ) 
             ),
             tabPanel(
               "Starting with Raw Data",
               value = "rawDataUiTab",
               IDBacApp::convertDataTab_UI("convertDataTab"))
             
  )
}
